src/batchoffer/BatchOffer.sol

```diff
@@ -0,0 +1,154 @@
+    marketConfig.checkIfCurrencyIsApproved(_currency);
+    require(_expiry > block.timestamp, "createBatchOffer::expiry must be in the future");
+    require(_amount > 0, "offer::Amount cannot be 0");
+
+    _creatorToRootToOffer[msg.sender][_rootHash] = BatchOffer(msg.sender, _rootHash, _amount, _currency, _expiry);
+
+    uint256 requiredAmount = _amount + marketConfig.marketplaceSettings.calculateMarketplaceFee(_amount);
+    MarketUtils.checkAmountAndTransfer(_currency, requiredAmount);
+
+    emit BatchOfferCreated(msg.sender, _rootHash, _amount, _currency, _expiry);
+  }
+
+  function revokeBatchOffer(bytes32 _rootHash) external nonReentrant() {
+    require(_creatorToRootToOffer[msg.sender][_rootHash].creator == msg.sender, "createBatchOffer::must be owner");
+
+    // Load Offer
+    BatchOffer memory offer = _creatorToRootToOffer[msg.sender][_rootHash];
+
+    // Cleanup memory
+    // IMPORTANT: Must be done before external refund call
+    delete _creatorToRootToOffer[msg.sender][_rootHash];
+
+    // Refund Escrow
+    uint256 fee = marketConfig.marketplaceSettings.calculateMarketplaceFee(offer.amount);
+    MarketUtils.refund(marketConfig, offer.currency, offer.amount, fee, offer.creator);
+  }
+
+  function acceptBatchOffer(
+    address _creator,
+    bytes32[] memory _proof,
+    bytes32 _rootHash,
+    address _contractAddress,
+    uint256 _tokenId,
+    address payable[] calldata _splitRecipients,
+    uint8[] calldata _splitRatios
+  ) external payable nonReentrant {
+    IERC721 erc721 = IERC721(_contractAddress);
+    address tokenOwner = erc721.ownerOf(_tokenId);
+
+    require(msg.sender == tokenOwner, "acceptBatchOffer::Must be tokenOwner");
+
+    BatchOffer memory offer = _creatorToRootToOffer[_creator][_rootHash];
+    address currency = offer.currency;
+    marketConfig.checkIfCurrencyIsApproved(currency);
+    require(offer.creator != address(0), "acceptBatchOffer::offer does not exist");
+    require(offer.expiry > block.timestamp, "acceptBatchOffer::offer expired");
+    require(offer.rootHash == _rootHash, "acceptBatchOffer::root mismatch");
+    bytes32 leaf = keccak256(abi.encodePacked(_contractAddress, _tokenId));
+    (bool success, ) = MerkleProof.verify(_proof, offer.rootHash, leaf);
+    require(success, "Invalid _proof");
+
+    // Cleanup memory
+    // IMPORTANT: Must be done before external payout
+    delete _creatorToRootToOffer[_creator][_rootHash];
+
+    // Perform payout
+    // TODO: test case to make sure mark as sold is performed after payout
+    if (offer.amount != 0) {
+      marketConfig.payout(
+        _contractAddress,
+        _tokenId,
+        currency,
+        offer.amount,
+        msg.sender,
+        _splitRecipients,
+        _splitRatios
+      );
+    }
+
+    // Transfer ERC721 token from the seller to the buyer
+    IERC721 erc721Token = IERC721(_contractAddress);
+    erc721Token.safeTransferFrom(msg.sender, offer.creator, _tokenId);
```

Comment:
might want to check if token is now owned by sender

Reply:
6108a096f32264c94994375fd0cd4e2ffb0f7046 do want to